File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.java
package nerd.tuxmobil.fahrplan.congress.schedule;

import android.content.Context;
import android.content.res.Resources;
import android.util.AttributeSet;
import android.view.GestureDetector;
import android.view.GestureDetector.SimpleOnGestureListener;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.HorizontalScrollView;
import android.widget.LinearLayout;

import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;

import info.metadude.android.eventfahrplan.commons.logging.Logging;
import nerd.tuxmobil.fahrplan.congress.R;


public class HorizontalSnapScrollView extends HorizontalScrollView {

    private static final String LOG_TAG = "HorizontalSnapScrollView";

    @NonNull
    private final Logging logging = Logging.get();

    private final GestureDetector gestureDetector;

    private int activeColumnIndex = 0;

    private int xStart;

    private int columnWidth;

    private int roomsCount = NOT_INITIALIZED;

    private HorizontalScrollView roomNames = null;

    private int displayColumnCount;

    private static final int NOT_INITIALIZED = Integer.MIN_VALUE;

    private static final int SWIPE_MIN_DISTANCE = 5;

    private static final int SWIPE_THRESHOLD_VELOCITY = 2800;

    /**
     * get currently displayed column index
     *
     * @return index (0..n)
     */
    public int getColumnIndex() {
        return activeColumnIndex;
    }

    public int getLastVisibleColumnIndex() {
        return activeColumnIndex + displayColumnCount - 1;
    }

    class YScrollDetector extends SimpleOnGestureListener {

        @Override
        public boolean onDown(MotionEvent e) {
            xStart = (int) e.getX();
//          logging.d(LOG_TAG, "onDown xStart: " + xStart + " getMeasuredWidth: " + getMeasuredWidth());
            float ofs = (float) (getScrollX() * displayColumnCount) / getMeasuredWidth();
            activeColumnIndex = Math.round(ofs);
            return super.onDown(e);
        }

        @Override
        public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
            float scale = getResources().getDisplayMetrics().density;
            float normalizedAbsoluteVelocityX = Math.abs(velocityX / scale);
            int columns = (int) Math.ceil((normalizedAbsoluteVelocityX / SWIPE_THRESHOLD_VELOCITY) * 3);
//	    	logging.d(LOG_TAG, "onFling " + velocityX + "/" + velocityY + " " + velocityX/scale + " " + columns);
            boolean exceedsVelocityThreshold = normalizedAbsoluteVelocityX > SWIPE_THRESHOLD_VELOCITY;
            try {
                //right to left
                if (e1.getX() - e2.getX() > SWIPE_MIN_DISTANCE && exceedsVelocityThreshold) {
                    scrollToColumn(activeColumnIndex + columns, false);
                    return true;
                }
                //left to right
                else if (e2.getX() - e1.getX() > SWIPE_MIN_DISTANCE && exceedsVelocityThreshold) {
                    scrollToColumn(activeColumnIndex - columns, false);
                    return true;
                }
            } catch (Exception e) {
                logging.e(LOG_TAG, "There was an error processing the Fling event:" + e.getMessage());
            }
            return super.onFling(e1, e2, velocityX, velocityY);
        }

    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return super.onTouchEvent(event);
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        boolean result = super.onInterceptTouchEvent(ev);
        gestureDetector.onTouchEvent(ev);
        return result;
    }

    /**
     * Sets the rooms count so it can be used to calculate
     * the room column dimensions in the next layout phase.
     */
    public void setRoomsCount(@IntRange(from = 1) int roomsCount) {
        this.roomsCount = roomsCount;
    }

    public void scrollToColumn(int col, boolean fast) {
        int max;
        if (columnWidth == 0) {
            max = 0;
        } else {
            max = (getChildAt(0).getMeasuredWidth() - columnWidth) / columnWidth;
        }
        if (col < 0) {
            col = 0;
        }
        if (col > max) {
            col = max;
        }
        final int scrollTo = col * columnWidth;
        logging.d(LOG_TAG, "scroll to col " + col + "/" + scrollTo + " " + getChildAt(0).getMeasuredWidth());
        if (!fast) {
            this.post(() -> smoothScrollTo(scrollTo, 0));
        } else {
            scrollTo(scrollTo, 0);
            if (roomNames != null) {
                roomNames.scrollTo(scrollTo, 0);
            }
        }
        activeColumnIndex = col;
    }

    public HorizontalSnapScrollView(Context context, AttributeSet attrs) {
        super(context, attrs);
        displayColumnCount = getResources().getInteger(R.integer.max_cols);
        columnWidth = 0;
        gestureDetector = new GestureDetector(new YScrollDetector());
        setOnTouchListener(new OnTouchListener());
    }

    private class OnTouchListener implements View.OnTouchListener {

        public boolean onTouch(View v, MotionEvent event) {
            logging.d(LOG_TAG, "onTouch");
            if (gestureDetector.onTouchEvent(event)) {
                // logging.d(LOG_TAG, "gesture detector consumed event");
                return false;
            } else if (event.getAction() == MotionEvent.ACTION_UP
                    || event.getAction() == MotionEvent.ACTION_CANCEL) {
                int scrollX = (int) event.getX();
                int distance = scrollX - xStart;
                logging.d(LOG_TAG,
                        "column width:" + columnWidth + " scrollX:" + scrollX + " distance:"
                                + distance + " active column index:" + activeColumnIndex);
                int columnIndex = activeColumnIndex;
                int columnDistance;
                if (displayColumnCount > 1) {
                    columnDistance = Math.round(Math.abs((float) distance) / columnWidth);
                    logging.d(LOG_TAG, "column distance: " + columnDistance);
                    if (distance > 0) {
                        columnIndex = activeColumnIndex - columnDistance;
                    } else {
                        columnIndex = activeColumnIndex + columnDistance;
                    }
                } else {
                    if (Math.abs(distance) > columnWidth / 4) {
                        if (distance > 0) {
                            columnIndex = activeColumnIndex - 1;
                        } else {
                            columnIndex = activeColumnIndex + 1;
                        }
                    }
                }
                scrollToColumn(columnIndex, false);

                return true;
            } else {
                return false;
            }
        }
    }

    /**
     * Calculates the number of columns to display at a time based on the physical dimensions and
     * the screen density of the device and the column count of the schedule. Further limiting
     * factors are a maximum column count to be displayed and a minimum column width.
     */
    @VisibleForTesting
    public static int calculateDisplayColumnCount(
            int availablePixels,
            int columnsCount,
            int maxColumnCountForLayout,
            float densityScaleFactor,
            int minColumnWidthDip
    ) {
        int columnCountLimit = Math.min(columnsCount, maxColumnCountForLayout);
        if (columnCountLimit == 1) {
            return 1;
        }
        float availableDips = ((float) availablePixels) / densityScaleFactor;
        int minWidthColumnCount = (int) Math.floor(availableDips / minColumnWidthDip);
        int columnCount = Math.min(minWidthColumnCount, columnCountLimit);
        return Math.max(1, columnCount);
    }

    @Override
    protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight) {
        logging.d(LOG_TAG, "onSizeChanged " + oldWidth + ", " + oldHeight + ", " + width + ", " + height + " getMW:" + getMeasuredWidth());
        super.onSizeChanged(width, height, oldWidth, oldHeight);
        Resources resources = getResources();
        if (roomsCount == NOT_INITIALIZED) {
            displayColumnCount = 1;
        } else {
            displayColumnCount = calculateDisplayColumnCount(
                    getMeasuredWidth(),
                    roomsCount,
                    resources.getInteger(R.integer.max_cols),
                    resources.getDisplayMetrics().density,
                    resources.getInteger(R.integer.min_width_dip));
        }

        int newItemWidth = Math.round((float) getMeasuredWidth() / displayColumnCount);
        float scale = getResources().getDisplayMetrics().density;

        logging.d(LOG_TAG, "item width: " + newItemWidth + " " + ((float) newItemWidth) / scale + "dp");
        setColumnWidth(newItemWidth);
    }

    @Override
    protected void onScrollChanged(int scrollX, int scrollY, int oldScrollX, int oldScrollY) {
        super.onScrollChanged(scrollX, scrollY, oldScrollX, oldScrollY);
//		logging.d(LOG_TAG, "scrolled from " + oldScrollX + " to " + scrollX);
        if (roomNames != null) {
            roomNames.scrollTo(scrollX, 0);
        }
    }

    public void setChildScroller(HorizontalScrollView h) {
        roomNames = h;
    }

    public int getColumnWidth() {
        if (roomsCount == NOT_INITIALIZED) {
            throw new IllegalStateException("The \"roomsCount\" field must be initialized before invoking \"getColumnWidth\".");
        }
        return columnWidth;
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        scrollToColumn(activeColumnIndex, true);
    }

    private void setColumnWidth(int pixels) {
        logging.d(LOG_TAG, "setColumnWidth " + pixels);
        columnWidth = pixels;
        if (pixels == 0) {
            return;
        }

        ViewGroup container = (ViewGroup) getChildAt(0);
        int childCount = container.getChildCount();
        for (int i = 0; i < childCount; i++) {
            ViewGroup c = (ViewGroup) container.getChildAt(i);
            ViewGroup.LayoutParams p = c.getLayoutParams();
            p.width = columnWidth;
            c.setLayoutParams(p);
        }
        container.measure(MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED),
                MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED));

        if (roomNames != null) {
            View v;
            int firstRoomChildCount = ((ViewGroup) roomNames.getChildAt(0)).getChildCount();

            for (int i = 0; i < firstRoomChildCount; i++) {
                v = ((ViewGroup) roomNames.getChildAt(0)).getChildAt(i);
                LinearLayout.LayoutParams p = (LinearLayout.LayoutParams) v.getLayoutParams();
                p.width = columnWidth;
                v.setLayoutParams(p);
            }
        }
        scrollToColumn(activeColumnIndex, true);
    }
}

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt
package nerd.tuxmobil.fahrplan.congress.schedule

import android.content.Context
import android.util.AttributeSet
import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.view.MotionEvent.ACTION_CANCEL
import android.view.MotionEvent.ACTION_UP
import android.view.View
import android.view.ViewGroup
import android.widget.HorizontalScrollView
import androidx.annotation.IntRange
import androidx.annotation.VisibleForTesting
import androidx.core.view.children
import androidx.core.view.forEach
import androidx.core.view.get
import androidx.core.view.updateLayoutParams
import info.metadude.android.eventfahrplan.commons.logging.Logging
import nerd.tuxmobil.fahrplan.congress.R
import nerd.tuxmobil.fahrplan.congress.repositories.AppRepository
import kotlin.Int.Companion.MAX_VALUE
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.roundToInt

class HorizontalSnapScrollView(context: Context, attrs: AttributeSet) : HorizontalScrollView(context, attrs) {

    companion object {
        private const val LOG_TAG = "HorizontalSnapScrollView"
        private const val FLING_COLUMN_MULTIPLIER = 3

        @VisibleForTesting
        const val SWIPE_DISTANCE_THRESHOLD = 5

        @VisibleForTesting
        const val SWIPE_VELOCITY_THRESHOLD = 2800

        @VisibleForTesting
        const val SWIPE_VELOCITY_THRESHOLD_UNREACHABLE = MAX_VALUE

        /**
         * Calculates the number of columns to display at a time based on the physical dimensions and
         * the screen density of the device and the column count of the schedule. Further limiting
         * factors are a maximum column count to be displayed and a minimum column width.
         */
        @VisibleForTesting
        fun calculateDisplayColumnCount(
            availablePixels: Int,
            columnsCount: Int,
            maxColumnCountForLayout: Int,
            densityScaleFactor: Float,
            minColumnWidthDip: Int
        ): Int {
            val columnCountLimit = columnsCount.coerceAtMost(maxColumnCountForLayout)
            if (columnCountLimit == 1) {
                return 1
            }
            val availableDips = availablePixels.toFloat() / densityScaleFactor
            val minWidthColumnCount = floor((availableDips / minColumnWidthDip).toDouble()).toInt()
            val columnCount = minWidthColumnCount.coerceAtMost(columnCountLimit)
            return 1.coerceAtLeast(columnCount)
        }
    }

    private val logging = Logging.get()
    private val gestureDetector: GestureDetector
    private var horizontalSnapScrollState = HorizontalSnapScrollState(logging)
    private val swipeThresholdVelocity
        get() = if (AppRepository.readFastSwipingEnabled()) SWIPE_VELOCITY_THRESHOLD else SWIPE_VELOCITY_THRESHOLD_UNREACHABLE
    private lateinit var roomNames: HorizontalScrollView

    /**
     * Get currently displayed column index
     * @return index (0..n)
     */
    val columnIndex: Int
        get() = horizontalSnapScrollState.activeColumnIndex

    init {
        horizontalSnapScrollState = horizontalSnapScrollState.copy(
            logging = horizontalSnapScrollState.logging,
            xStart = horizontalSnapScrollState.xStart,
            displayColumnCount = resources.getInteger(R.integer.max_cols),
            roomsCount = horizontalSnapScrollState.roomsCount,
            columnWidth = 0,
            activeColumnIndex = horizontalSnapScrollState.activeColumnIndex
        )
        overScrollMode = OVER_SCROLL_NEVER
        gestureDetector = GestureDetector(context, YScrollDetector())
        setOnTouchListener(OnTouchListener())
    }

    internal inner class YScrollDetector : SimpleOnGestureListener() {
        override fun onDown(motionEvent: MotionEvent): Boolean {
            val xStart = motionEvent.x.toInt()
            logging.d(LOG_TAG, "onDown -> xStart: $xStart, measuredWidth: $measuredWidth")
            val ofs = (scrollX * horizontalSnapScrollState.displayColumnCount).toFloat() / measuredWidth
            horizontalSnapScrollState = horizontalSnapScrollState.copy(
                logging = horizontalSnapScrollState.logging,
                xStart = xStart,
                displayColumnCount = horizontalSnapScrollState.displayColumnCount,
                roomsCount = horizontalSnapScrollState.roomsCount,
                columnWidth = horizontalSnapScrollState.columnWidth,
                activeColumnIndex = ofs.roundToInt()
            )
            return super.onDown(motionEvent)
        }

        override fun onFling(
            start: MotionEvent?,
            end: MotionEvent,
            velocityX: Float,
            velocityY: Float
        ): Boolean {
            if (start != null) {
                val normalizedVelocityX = velocityX / resources.displayMetrics.density
                val columns = ceil((normalizedVelocityX / swipeThresholdVelocity * FLING_COLUMN_MULTIPLIER).toDouble()).toInt()

                logging.d(LOG_TAG, "onFling -> $velocityX/$velocityY $normalizedVelocityX $columns")

                if (isLongEnough(start = start.x, end = end.x, SWIPE_DISTANCE_THRESHOLD) && isFastEnough(velocity = normalizedVelocityX, swipeThresholdVelocity)) {
                    scrollToColumn(horizontalSnapScrollState.activeColumnIndex - columns, fast = false)
                    return true
                }
            }
            return super.onFling(start, end, velocityX, velocityY)
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        return super.onTouchEvent(event)
    }

    override fun onInterceptTouchEvent(event: MotionEvent): Boolean {
        gestureDetector.onTouchEvent(event)
        return super.onInterceptTouchEvent(event)
    }

    /**
     * Sets the rooms count so it can be used to calculate
     * the room column dimensions in the next layout phase.
     */
    fun setRoomsCount(@IntRange(from = 1) roomsCount: Int) {
        horizontalSnapScrollState = horizontalSnapScrollState.copy(
            logging = horizontalSnapScrollState.logging,
            xStart = horizontalSnapScrollState.xStart,
            displayColumnCount = horizontalSnapScrollState.displayColumnCount,
            roomsCount = roomsCount,
            columnWidth = horizontalSnapScrollState.columnWidth,
            activeColumnIndex = horizontalSnapScrollState.activeColumnIndex
        )
    }

    fun scrollToColumn(col: Int, fast: Boolean) {
        val scrollTo = horizontalSnapScrollState.calculateScrollToXPosition(col)
        logging.d(LOG_TAG, "scroll to col $col/$scrollTo ${getChildAt(0).measuredWidth}")
        if (fast) {
            scrollTo(scrollTo, 0)
            roomNames.scrollTo(scrollTo, 0)
        } else {
            post { smoothScrollTo(scrollTo, 0) }
        }
        horizontalSnapScrollState = horizontalSnapScrollState.copy(
            logging = horizontalSnapScrollState.logging,
            xStart = horizontalSnapScrollState.xStart,
            displayColumnCount = horizontalSnapScrollState.displayColumnCount,
            roomsCount = horizontalSnapScrollState.roomsCount,
            columnWidth = horizontalSnapScrollState.columnWidth,
            activeColumnIndex = col
        )
    }

    private inner class OnTouchListener : View.OnTouchListener {
        override fun onTouch(view: View, event: MotionEvent): Boolean {
            logging.d(LOG_TAG, "onTouch")
            if (gestureDetector.onTouchEvent(event)) {
                // logging.d(LOG_TAG, "gesture detector consumed event")
                return false
            } else if (event.action == ACTION_UP || event.action == ACTION_CANCEL) {
                val scrollX = event.x.toInt()

                val calculatedColumnIndex = horizontalSnapScrollState.calculateOnTouchColumnIndex(scrollX)

                if (calculatedColumnIndex > -1) {
                    scrollToColumn(calculatedColumnIndex, fast = false)
                    return true
                }
            }
            return false
        }
    }

    override fun onSizeChanged(width: Int, height: Int, oldWidth: Int, oldHeight: Int) {
        logging.d(LOG_TAG, "onSizeChanged -> $oldWidth, $oldHeight, $width, $height, measuredWidth: $measuredWidth")
        super.onSizeChanged(width, height, oldWidth, oldHeight)
        val displayColumnCount = if (horizontalSnapScrollState.isRoomsCountInitialized()) {
            calculateDisplayColumnCount(
                availablePixels = measuredWidth,
                columnsCount = horizontalSnapScrollState.roomsCount,
                maxColumnCountForLayout = resources.getInteger(R.integer.max_cols),
                densityScaleFactor = resources.displayMetrics.density,
                minColumnWidthDip = resources.getInteger(R.integer.min_width_dip)
            )
        } else {
            1
        }
        horizontalSnapScrollState = horizontalSnapScrollState.copy(
            logging = horizontalSnapScrollState.logging,
            xStart = horizontalSnapScrollState.xStart,
            displayColumnCount = displayColumnCount,
            roomsCount = horizontalSnapScrollState.roomsCount,
            columnWidth = horizontalSnapScrollState.columnWidth,
            activeColumnIndex = horizontalSnapScrollState.activeColumnIndex
        )
        val newItemWidth = horizontalSnapScrollState.calculateColumnWidth(measuredWidth)
        logging.d(LOG_TAG, """item width: $newItemWidth ${newItemWidth.toFloat() / resources.displayMetrics.density}dp""")
        columnWidth = newItemWidth
    }

    override fun onScrollChanged(scrollX: Int, scrollY: Int, oldScrollX: Int, oldScrollY: Int) {
        super.onScrollChanged(scrollX, scrollY, oldScrollX, oldScrollY)
        // logging.d(LOG_TAG, "onScrollChanged -> scrolled from $oldScrollX to $scrollX")
        roomNames.scrollTo(scrollX, 0)
    }

    fun setChildScroller(titleScrollView: HorizontalScrollView) {
        roomNames = titleScrollView
    }

    var columnWidth: Int
        get() {
            check(horizontalSnapScrollState.isRoomsCountInitialized()) {
                """The "roomsCount" field must be initialized before invoking "getColumnWidth"."""
            }
            return horizontalSnapScrollState.columnWidth
        }
        private set(pixels) {
            logging.d(LOG_TAG, "setColumnWidth: $pixels")
            horizontalSnapScrollState = horizontalSnapScrollState.copy(
                logging = horizontalSnapScrollState.logging,
                xStart = horizontalSnapScrollState.xStart,
                displayColumnCount = horizontalSnapScrollState.displayColumnCount,
                roomsCount = horizontalSnapScrollState.roomsCount,
                columnWidth = pixels,
                activeColumnIndex = horizontalSnapScrollState.activeColumnIndex
            )

            if (pixels == 0) {
                return
            }

            val container = firstChild
            container.forEach {
                it.updateLayoutParams {
                    width = horizontalSnapScrollState.columnWidth
                }
            }
            container.measure(
                MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED),
                MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED)
            )

            roomNames.firstChild.children.forEach {
                it.updateLayoutParams {
                    width = horizontalSnapScrollState.columnWidth
                }
            }

            scrollToColumn(horizontalSnapScrollState.activeColumnIndex, fast = true)
        }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        scrollToColumn(horizontalSnapScrollState.activeColumnIndex, fast = true)
    }
}

private val ViewGroup.firstChild
    get() = get(0) as ViewGroup

/**
 * Returns a boolean indicating if the velocity is fast enough for a fling-scroll.
 */
@VisibleForTesting
fun isFastEnough(velocity: Float, threshold: Int) =
    abs(velocity) > threshold

/**
 * Returns a boolean indicating if the absolute distance between two values is big enough for a fling-scroll.
 */
@VisibleForTesting
fun isLongEnough(start: Float, end: Float, threshold: Int) =
    abs(start - end) > threshold
